### DISCUSSION TITLE
[Proposal]: Make unit testing in C# simpler for mocking classes without making members virtual

### DISCUSSION THREAD

**Discussion by ChiragRupani** [dotnet/csharplang#6251] [Language Ideas]
**TLDR**:
Proposal to provide framework / language support to mock members for unit testing without declaring it as virtual.

**Current Scenario:**
Similar to [Sealed class peanut butter](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-6/#peanut-butter), I would like to declare class as `sealed`. There are many cases where class needs to be unsealed so that it can be mocked and requires method be declared `virtual` or be extracted to interface even if there is no need of public interface contracts to be shared e.g. in ASP.NET Core API where Swagger doc is shared . This is only done for unit / application testing - which seems quite odd as just for testing we are opening class which otherwise shouldn't be be case. It also breaks encapsulation and may or [may not](https://devblogs.microsoft.com/dotnet/performance-improvements-in-ryujit-in-net-core-and-net-framework/) affect performance as well.

Currently, the mocking classes usually services that makes network call (like to database or other API) requires the method to be declared as `virtual`. Most of current mocking libraries are based on Castle Dynamic Proxy or similar. The general idea is it proxies either implements interface or derives from class and intercept the invocations and direct to mock code. However, this overriding behavior is only possible if the method is declared as `virtual` (either member declared directly as virtual in class or with implementing the interface). 

**Proposal:**
There should be built-in support in framework to mock code where it is used without making it virtual. 

One way can be following - Though `InternalsVisibleTo` attribute applied using Manifest, the similar attribute may be defined for this. In that case the extra step can be added to generate IL with transformation - as if all members are declared as virtual and class as without sealed keyword and embedding the result assembly in test project instead of actual assembly which is generated without any transformation. The test project should ignore referenced actual assembly. This is just thought, there must be better / elegant to address this.

Anything that doesn't require to make changes in original code (like making it virtual), just to make it usable in test code would be helpful. Hope this is considered for further discussion.

**Comment by yaakov-h** [comment]
If you're that desperate to have a class that is sealed but not:

```csharp
public
#if !DEBUG
sealed
#endif
class
```

**Reply by ChiragRupani** [reply]
Not desperate, but would you write such code for every class in actual production code?

**Reply by yaakov-h** [reply]
I wouldn't write my production code in such a way that I need to unseal it in order to test it.

**Comment by TonyValenti** [comment]
Take a look at LibHarmony. It lets you override non-virtual members using some amazing magic.

**Reply by ChiragRupani** [reply]
Better there are many IL weavers like plugins based on Fody. But I want this functionality baked in language / framework

**Comment by theunrepentantgeek** [comment]
How do you envisage the compiler injecting _seams_ into the objects for testing purposes without using the existing mechanisms (virtual methods and interfaces) that allow such _seams_ to be introduced?

> as if all members are declared as virtual and class as without sealed keyword and embedding the result assembly in test project instead of actual assembly which is generated without any transformation

It's straightforward to identify scenarios where copying the type into another assembly will just break. Remember that types in different assemblies are _different types_, even if they have the same name. 

As a starting point for pondering, consider any reasonable override of `Equals(object)`, which almost inevitably starts with a type test to see if the passed object is of an expected type. Don't forget that `a.Equals(b)` has to give the same result as `b.Equals(a)`, else you run into all sorts of pain with dictionaries, hashmaps, lists, and the like.

> just for testing we are opening class which otherwise shouldn't be be case

This is why many people advocate for testing only against the external interface of the type - and why many of those same people believe that smaller types are better. (For the record, I'm not in the first camp, but I am firmly in the second.)

**Comment by Kebechet** [comment]
Worth revisiting this thread now that **interceptors** have moved past the experimental stage:

- Compiler-stable since .NET 9.0.2xx SDK ([roslyn/docs/features/interceptors.md](https://github.com/dotnet/roslyn/blob/main/docs/features/interceptors.md))
- Opt-in via `<InterceptorsNamespaces>` in csproj — no preview flag needed
- Already used in production by ASP.NET Core's RDG / Minimal APIs source generators
- Language-spec ratification still pending ([#7009](https://github.com/dotnet/csharplang/issues/7009) is "Implemented Needs ECMA Spec")

Interceptors look like a near-perfect fit for the problem this discussion was originally raised about. A mocking library could ship a Roslyn analyzer that, given:

```csharp
Mock.Replace(() => realService.DoWork(arg), returns: 42);
```

emits an `[InterceptsLocation]` redirect at the call site to a generated stub. No interface, no `virtual`, no profiler-based runtime instrumentation, no IL rewriting. Compile-time only, AOT-friendly, debugger-friendly.

Yet — as far as I can find — **no mainstream mocking library has shipped or even publicly prototyped this**. Moq, NSubstitute, FakeItEasy all still require interfaces or `virtual`. The closest thing in the Moq camp ([moq/labs#95](https://github.com/moq/labs/discussions/95)) goes a different direction (Moq.Sdk / Avatar) and never references `[InterceptsLocation]`.

My read: library authors are hesitant because interceptors aren't yet sanctioned as a *language feature* (still missing from the "What's new in C# 14" and "What's new in C# 15" pages), so taking a hard dependency feels risky.

Two questions for the community / LDM:

1. **Are non-framework-author scenarios like mocking explicitly in scope** for the eventual spec ratification, or is the intent to keep interceptors as a framework-only primitive? If it's the latter, that's worth saying out loud so the mocking ecosystem stops waiting.
2. **If they are in scope** — is anyone aware of a public prototype? This seems like the kind of thing where one weekend hack from a Moq/NSubstitute maintainer would unblock a huge ergonomic win for the whole .NET testing story.

**Reply by CyrusNajmabadi** [reply]
>  or is the intent to keep interceptors as a framework-only primitive? 

there is no intent for interceptors to be framework only.  it's a general roslyn extension point, just like analyzers and generators.

**Reply by jnm2** [reply]
Roslyn interceptors cannot intercept calls that are in compiled binaries, which includes call sites in the SUT which is defined outside the test project.

**Reply by Kebechet** [reply]
Then what would be the proper solution for this?

**Reply by ChiragRupani** [reply]
By adding Mock Generator as OutputItemType = Analyzer, can't it intercept call?

```xml
 <ItemGroup>
   <ProjectReference Include="..\MockGenerator\Mock.Generator.csproj"  OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
 </ItemGroup>
 ```

or package
```
  <ItemGroup>
    <PackageReference Include="Mock.MockGenerator" Version="1.0.0"  />
  </ItemGroup>
```
 
 Where MockGenerator has following in csproj file:
 ```xml
 <IsRoslynComponent>true</IsRoslynComponent>
 <IncludeBuildOutput>true</IncludeBuildOutput>
 <EnforceExtendedAnalyzerRules>true</EnforceExtendedAnalyzerRules>
```

**Reply by toupswork** [reply]
The C# language team likely never prioritized this because they see unit test mocking as a tooling issue, and for a while it was in the Enterprise version of Visual Studio. But those testing utilities have gone cold. 
Not everything has a mocking workaround. Others have use IL Weaving or removed the sealed attribute or added partial to the assembly method metadata. It would just be nice if there was a language first way to do this.

**Reply by jnm2** [reply]
@ChiragRupani A Roslyn interceptor can only intercept direct calls that are made within the source code of the project that references the analyzer. If the non-test project has:
```cs
class ClassDefinedInOtherProject
{
    public void A() => Console.WriteLine("Not intercepted");
    public void B() => A();
}
```
And your unit test project has:
```cs
var sut = new ClassDefinedInOtherProject();
sut.A();
sut.B();
```
And the unit test project references a Roslyn interceptor which intercepts calls to ClassDefinedInOtherProject.A, it can intercept the `sut.A()` line and redirect it do something else, but it can not intercept the `A()` call which is made inside `ClassDefinedInOtherProject.B`.

**Reply by jnm2** [reply]
And even if you add a Roslyn interceptor to the project that declares ClassDefinedInOtherProject, you will still have scenarios that can't be intercepted. For example, imagine:
```cs
class ClassInProjectReferencingInterceptor : ICloneable
{
    public object Clone() => ...
}
```
And outside the current project, someone does:
```cs
void M(IClonable clonable)
{
    var x = cloneable.Clone();
}
```
A Roslyn interceptor can do nothing here to intercept the call to `ClassInProjectReferencingInterceptor.Clone()` that is performed via the interface call.

The runtime already has a function for this, which is virtual calls. That is the hook to use. With IL rewriting you can make the method virtual to do the interception, and perhaps with fancy debugger APIs you can hook the method's execution and inject test code, but those are fantastically complicated compared to just making the method virtual. The failure modes around these approaches waste a lot of time compared to just using the runtime feature and making the method virtual.

_In general_ it's an architecture smell if a test wants to intercept a method that's not happily and easily a virtual method or an interface method. Tests should treat the system under test as a black box and should not have special access to internal state.

**Reply by CyrusNajmabadi** [reply]
> The C# language team likely never prioritized this because they see unit test mocking as a tooling issue

That's not true.  

>  It would just be nice if there was a language first way to do this.

There *are* language first ways to do things.  The issue is that people explicitly avoid them, and then ask "what language solution is there for this, given that i will not use this existing language solution".  But that is not tenable for us to address.  Because, no matter what we then end up adding, someone will say "i will not use that, what is your new language solution to address me not using any of the available mechanisms to do this..." :)

**Reply by toupswork** [reply]
OK, here's a real-life example. I had a 3rd party library that had a class I had to use for authentication and there was code in the constructor that would throw in unit tests. The class was sealed and has no interface. What is the language-first way to mock that? No wrapping it doesn't work. 

I bypassed the ctor by using GetUninitialiedObject() in RuntimeHelpers, which is not exactly a meant for testing. I have been a C# developer since .NET 1.1. So I happen to know how to force my way through things if I need to. You used to have Microsoft Fakes and Stubs but then they have been abandoned and no longer work.

**Reply by CyrusNajmabadi** [reply]
> Stop acting like everyone else is dumb.

This is not constructive, and violate the acceptable standards of our code of conduct.

**Reply by CyrusNajmabadi** [reply]
> The other day I posted a question about closed types and virtualization, and you were so eager to treat the question like it was asked by an idiot that it led to your embarrassment when you tried to quickly close it thinking I was asking about union types and how union types are not inherited.

What on earth are you talking about?

My first response was an apology about me misreading**.  I treated it appropriately based on how i had misread it.  And then when i realized mistake after your response, immediately answered appropriately with that clarification.  At no point did i think or act as if it was being asked by an idiot.

Again, this behavior is not constructive.  People are accurately and appropriately answering questions based on their assessments of the scenarios and situations here.  

>  I remember every feature added to C# in the last 25 years because I lived through it.

So have many of us :). I worked on C# itself from 1.0 onwards :) 

--

** This was my response btw:

> Very sorry. Deep brain fart reading the email.
>
> In terms of closed there's nothing stopping this from being utilized by the runtime. It's entirely up to them if that's something they'd find useful to optimize. The language (designed here) doesn't care one way or the other. We design the semantics and execution model of the lang. Optimization work (esp. low level from a particular runtime) is an orthogonal/external concern.
> 
> And, similar to how things like async/await work, it's entirely possible for a future direct runtime investment that the compiler (like Roslyn) can then take advantage of. Again, it's outside the lang's concern. Like with async/await, we just define the semantics. And later efforts can transparently make things faster. Especially in a way that allows users to run just find on existing runtimes, or get better over time as the compilers improve.

This response started clearly with an apology for my mistake.  And i then answered appropriately and considerately based on my experience actually working on the design of this feature.

**Reply by CyrusNajmabadi** [reply]
> So tell me what language features you all have added to C# to aid in unit tests. 

My professional opinion is that virtual is one of the most ideal aids for substituting test behaviors.  

Alternatives include abstracting behind interfaces with different implementations (including standard impls in non-test environs, and test impls in test environs). 

If the appropriate mechanisms are avoided, then there's really nothing the lang can do to help.  Because any solution we provide can simply be added by the next customer to the list of things they will not use.  And philosophically, we are opposed to adding alternatives just because someone doesn't like the preexisting solution.

**Reply by toupswork** [reply]
Thank you for that.  I am going to move on now from this discussion. If I took you wrong, my mistake. I'll let you all carry on with the discussion. Have a nice day. 😀

**Reply by jnm2** [reply]
> OK, here's a real-life example. I had a 3rd party library that had a class I had to use for authentication and there was code in the constructor that would throw in unit tests. The class was sealed and has no interface. What is the language-first way to mock that? No wrapping it doesn't work.

Wrapping some boundary is the way to go. The boundary may be larger than "constructor call." A unit test would generally be mocking out the entire part of the 3rd party library interaction that hits an external system. An integration test would generally be providing enough configuration so that some external system is hit without introducing new hooks that the 3rd party authentication layer doesn't itself provide.

**Comment by toupswork** [comment]
This is something that has bugged me for so many years that I ended up writing a utility that redirects methods (virtual, generic, static, sealed, instance) by patching the function preamble to perform a long jump. It's ridiculous. Of course, there is always IL Weaving or patching the assembly to remove "sealed" from a class or add modifiers to functions like "virtual"

**Comment by jnm2** [comment]
If you wouldn't want a method to be overridden in production, that is how you know that it should not be overridden in a unit test. There are cleaner ways to approach the unit testing, with better separation of responsibilities between classes in production, so that there is no desire to mock a nonvirtual method.

**Reply by BlinD-HuNTeR** [reply]
I agree that methods shouldn't be marked virtual, just because of unit testing. The final goal of all code is to be run in production, and if there is no polymorphism involved (and no expectation of polymorphism in the future), then it shouldn't be virtual.

Nevertheless, I still think the .NET ecosystem should expose some way of intercepting any piece of code, in a development environment. Interceptors are currently the best alternative, but they are compile-time only, which means that, if the code being tested uses 3rd-party libraries, there's no way to mock anything that is invoked by those libraries, you would have to mock the entire 3rd-party library itself.

On top of that, it all comes down to usability: devs will use what is available to them, and given that all mocking libraries are working on top of virtual methods, that's what everybody is doing currently.

Interceptors are powerful, but hard to use, coding for the Roslyn API is not trivial, and in the context of mocking, it's even worse: all currently existing mocking solutions expect you to say "what is being mocked", but interceptors, on the other hand, require you to tell "the specific method call that you want to intercept". I believe this restriction makes it very impractical for a mocking framework to ship a source generator that emits interceptors, and it's no surprise, that none of them did.

And look, I'm not advocating that this restriction be removed, I think it would be impractical to introduce a language feature that intercepts all calls of a given method, instead of specific ones. I'm just saying that, there should be a simple and easy way to intercept anything we want: it could be based on some existing profiling functionality, or maybe some new API that replaces a method at runtime, but only if some specific environment variable is set.

And as an honorable mention, I remember that once I found [this](https://www.nuget.org/packages/Pose) crazy and convoluted solution that uses Dynamic Methods: it would take a delegate, and emit a Dynamic Method with the same body as that delegate, but while doing that, it would take all methods that are invoked by that body, and emit duplicates of them too, doing so recursively, until it finds the thing you want to mock, and then it emits a call to the mocking code instead of the original method. Effectively duplicating all the code at runtime, which is a bit buggy and definitely not very performant, but it achieves the goal of intercepting anything.

**Reply by CyrusNajmabadi** [reply]
> The final goal of all code is to be run in production

> in a development environment.

one man's 'production' is another's "development environment" :)

**Reply by CyrusNajmabadi** [reply]
> or maybe some new API that replaces a method at runtime, but only if some specific environment variable is set.

this feels like an api to ask the runtime people then.  

--

The main issue i have is that existing solutions work so well, that i've never even felt there's a pain point justifying a need for this sort of thing.

**Reply by BlinD-HuNTeR** [reply]
> one man's 'production' is another's "development environment" :)

That's true 🙂. Still, the underlying message is that, we don't want it be possible to intercept any method at runtime, with no restriction. It has to have some way to be restricted to specific environments, maybe controlled by something like a `DOTNET_` environment variable.

> existing solutions work so well

Existing solutions have these annoying limitations, where you can't mock 3rd-party code that is non-virtual, and if it is your own code, then you have to make it `virtual` when it shouldn't be.

I do agree, however, that this discussion is not relevant for [csharplang](https://github.com/dotnet/csharplang). It is more relevant for runtime and/or the developers of mocking libraries.

**Reply by CyrusNajmabadi** [reply]
> maybe controlled by something like a DOTNET_ environment variable.

What if someone ends up needing to intercept in that environment to address a problem found in production? :)

This really doesn't seem like people are asking for testing.  This sort of "i want to intercept everything" capability is effectively: give me control over the runtime.  Regardless of how you slice/dice it, this is firmly in their camp.

> where you can't mock 3rd-party code that is non-virtual,

I dunno, i use 3rd party code all the time, and i test just fine with it.  I use bog standard patterns which i think are good ideas *anyways*.  Namely, already hiding those concrete impls behind things like interfaces.  That way i'm not locked into their apis in the first place.  I can trivially swap out, override, etc.  These are simple patterns that have worked well since *v1*.  They're so ubiquitous and normal throughout *all* ecosystems.  So i'm very wary to try to come up with someting that is likely so very complex, when the existing solutions are easy, work great, and are readily available everywhere.

**Reply by Kebechet** [reply]
I think that is also my understanding...not just not using `virtual` but I dont want to change my code to be "testable" (virtual, interfaces etc) but still I want to be able to test (unit test in this context) all aspects of it.

Because most times what I saw was just hide everything behind interfaces so u can mock it...because of that u dont use interfaces to force some bondary on sibling classes etc...but just for the sake of testing.

So Idk if interceptors should be the way to go, but we need 1st class feature to resolve this somehow.

<img width="558" height="549" alt="images" src="https://github.com/user-attachments/assets/962b207b-9611-4605-a53b-feab3dac590a" />

**Reply by HaloFour** [reply]
Here's an open source library that claims to be able to intercept arbitrary members (including static members) for the sake of testing.  Doesn't look like it's been actively maintained for a while, though.

https://github.com/tonerdo/pose

I'm with Cyrus, tho.  This is a runtime concern, not a language one.  Java solves this by making it _relatively_ easier to insert custom instrumentation and class loaders which can transform classes as they are loaded, so the open source ecosystem around call interception is fairly robust and well supported.

**Reply by yaakov-h** [reply]
> some way of intercepting any piece of code

I have used https://github.com/pardeike/harmony in some exceptional unit tests for this.

**Reply by BlinD-HuNTeR** [reply]
> Here's an open source library that claims to be able to intercept arbitrary members (including static members) for the sake of testing. Doesn't look like it's been actively maintained for a while, though.
> 
> https://github.com/tonerdo/pose

I mentioned that in my post too. Have tested it in the past, and it's pretty crazy. It works by duplicating all code at runtime, recursively emitting dynamic methods that are copies of the methods to be executed, while replacing specific calls with calls to mocks.

> I have used https://github.com/pardeike/harmony in some exceptional unit tests for this.

Libraries like MonoMod.RuntimeDetour and Harmony (which also uses MonoMod internally), work by performing native hooking: they will take the memory address of the generated JIT code for a given method, and replace it with a trampoline.

I personally don't like these kinds of solutions, and I refrain from using them in my UnlimitedSharp project. First, because you are leaving the managed land and entering the realm of native code, which requires dealing with architecture-specific instructions (although these libraries already have support for ARM, so you may not see a problem on that). And second, which is my main concern: modern versions of .NET have tiered JIT, which means that a method may be re-jitted AFTER hooks are installed.

And to deal with that, MonoMod hooks the JIT interface, but that's an undocumented API, which CHANGES for every version of .NET. This effectively means that, if your code uses MonoMod, it loses forward compatibility: it will not work on future versions of the .NET runtime. Which for me, is unacceptable for production use, but may be fine for testing scenarios (you will still have to update MonoMod every time you upgrade to a new .NET runtime).

>class loaders which can transform classes as they are loaded

I really like this approach, and I believe it could be replicated in .NET as well. It should be possible to use the `AssemblyResolve` event to intercept assembly loading, then load the bytes of the target assembly in memory, and modify them with something like Cecil or dnLib.

However, AFAIK, `AssemblyResolve` only gets called when the rumtime **fails** to find an assembly in the disk. And so, I am going to drift away from a philosophical discussion, into a technical question: Is there a way to intercept assembly loading **before** the runtime tries to locate an assembly in the disk? Of course I am assuming that my loader will execute before everything else, which can be done with startup hooks.

**Reply by jnm2** [reply]
> Because most times what I saw was just hide everything behind interfaces so u can mock it...because of that u dont use interfaces to force some bondary on sibling classes etc...but just for the sake of testing.

Interface-per-class is generally an undesirable extreme. Coupling directly to external systems is the opposite, equally undesirable, extreme. There's a happy medium which is a fruitful space, where you introduce interfaces to _document and organize_ a small number of important boundaries. This is not done in a knee-jerk fashion. It can be done just-in-time, when the testing shows the unwanted coupling.

**Reply by AartBluestoke** [reply]
> "[CyrusNajmabadi](https://github.com/CyrusNajmabadi)
> ...

> "My professional opinion is that virtual is one of the most ideal aids for substituting test behaviors.
> 
> Alternatives include abstracting behind interfaces with different implementations (including standard impls in non-test environs, and test impls in test environs).
> 
> If the appropriate mechanisms are avoided, then there's really nothing the lang can do to help. 
> There are language first ways to do things. The issue is that people explicitly avoid them, and then ask "what language solution is there for this, given that i will not use this existing language solution". But that is not tenable for us to address. Because, no matter what we then end up adding, someone will say "i will not use that, what is your new language solution to address me not using any of the available mechanisms to do this..." :)"
> 

so it is the language teams opinion that given a third party library that expose import but currently non-virtual calls either:

- the library should re-architecture (potentially by just adding virtual to all functions relevant for any testing scenario, or exposing otherwise the appropriate injectable virtual call), 
- the libraries involved in any test scenario should be wrapped in an interface rather than directly used, 
- libraries are considered by the language team as calls that should not be on a testing path as their behavior is not supposed to be mocked for testing, 
- developers need to be rearchitecture their code splitting the along the unit testing lines that the language team consider to be reasonable boundaries, adding the appropriate abstractions of any non-mockable third party call?
- that doing anything other than the above is a pure tooling problem?

edit: @CyrusNajmabadi  - thanks for the follow-up clarification (edited the thanks in here, rather than another post, to avoid another ping)

**Reply by CyrusNajmabadi** [reply]
Yes.  I would either see if that library is willing to make their code more mockable.  And if they won't (or in the meantime), simply define your expected boundary api, and implement that on top of these libraries (or your own mocks). 

I've been working in many ecosystems for decades, and this approach has worked exceptionally well in all of them.
